### DISCUSSION TITLE
conversion of rate desired necessary?

### DISCUSSION THREAD

**Discussion by HelloIDTC** [bitcraze/discussions#1303] [Q&A]
Whether the rateDesired.roll, rateDesired.pitch, rateDesired.yaw needs to be converted to the body coordinate system before executing the attitudeControllerCorrectRatePID(sensors->gyro.x, -sensors->gyro.y, sensors->gyro.z, rateDesired.roll, rateDesired.pitch, rateDesired.yaw); ?
Because the sensors->gyro.x, gyro.y, gyro.z is measured in the body coordinate system, roll, pitch, yaw euler's angular (rate) is in Earth coordinates.

**Comment by knmcguire** [comment]
This doesn't seem to be happening in the function: https://github.com/bitcraze/crazyflie-firmware/blob/c2ddfe56cd1016e3182088eba8d7799a3d4c94ca/src/modules/src/controller/attitude_pid_controller.c#L141

I guess it seems to work as rate is relative, and as long as the changes are small it kind of works? But yes with a quadcopter rates tend to be much higher so that is perhaps not a great assumption. However so far it has been working out in practise but if we head over to faster platforms like the brushless we might need to keep this in mind

**Reply by HelloIDTC** [reply]
Why does the Ardupilot need to convert?
This function completes the coordinate system conversion function (update_rate_contoller_targets - converts earth frame rates to body frame rates for rate controllers):
https://github.com/ArduPilot/ardupilot/blob/ArduCopter-2.9-release/ArduCopter/Attitude.pde#L251

**Comment by tobbeanton** [comment]
> Roll, pitch, yaw euler's angular (rate) is in Earth coordinates.

What do you mean by this? Do you mean relative to the Earth coordinates?

**Reply by HelloIDTC** [reply]
I mean is that the roll, pitch, yaw (rate) is relative to the Earth coordinate system, but the angular rate measured by the gyroscope is relative to the body coordinate system. So we need to convert the coordinate system.

**Reply by gemenerik** [reply]
Check out our [coordinate system here](https://www.bitcraze.io/documentation/system/platform/cf2-coordinate-system/). The attitude angles are in the body frame.
